### ISSUE TITLE
Make Run.Files an array

### ISSUE THREAD

**Issue by vScottLouvau** [oasis-tcs/sarif-spec#256]
In working on the DeferredReader and binary format, I've come to a painful conclusion. :/

The Sarif Run files collection should be an array, not a dictionary, and properties referring to files should refer to them by integer index in the files array.

In both the DeferredReader and the binary format, very little information is required to "seek" into a big array quickly - store the positions of all or some array entries, seek to the nearest one, and walk forward. There's no equivalent way to provide seek into a dictionary; you must store every key and the position of the value in order to find values without walking everything.

This is why the DeferredReader used 33MB in the 400MB of FileData case but under 4MB in the 450MB of Results case - the seek data for the FileData collection is much larger.

This change would prevent Sarif producers from using the File URI as the file key. It's the obvious key choice, but means that long file URIs are often repeated several times in the data (for the file key, the FileData.Uri, and the references to it from elsewhere). 

In my 400MB FileData sample, which is a real credential scan output on a very large repo, writing files as an array reduces the file size by 36%.

Sarif producers who produce results as they scan files can simply add a new FileData to the Run.Files array and use (Run.Files.Count - 1) as the key they use to refer to it as they generate Results. Sarif producers who keep a whole graph of files in mind (perhaps when generating call stacks) would need to keep a dictionary, but they likely already require one and have committed to allocating enough memory to keep a "more than one file" model for analysis.

This change would not break the ability of consumers to quickly retrieve the FileData when referenced.
They would simply use Run.Files[Result.Locations[0].PhysicalLocation.Id] instead of Run.Files[Result.Locations[0].PhysicalLocation.Uri].

This would definitely be a substantial breaking change. I'm only proposing it because it would make SARIF files distinctly smaller and enable us to offer partial loading and fast search with much lower memory costs.

**Comment by vScottLouvau** [comment]
I think the File object contains the Uri and UriBaseId, as normal, but the PhysicalLocation object contains EITHER a Uri and optional UriBaseId OR an "id". If an "id" is present, the PhysicalLocation refers to the same file as Run.Files[Id], and the Uri and UriBaseId there should be considered the Uri for the location.

So:
"files": [
    { "uri": "io/stream.c", "uriBaseId": "SRCROOT" },
    ...
],
...
"results": [
    { 
       "ruleId": "CS0001", 
       "locations": [ 
            { "physicalLocation": {
                  "id": 0,   // Run.Files[0] provides 
                  "region": { "startLine": 24, "startColumn": 9 }
            }
        ]
    }
    ...
]

**Comment by michaelcfanning** [comment]
I've been thinking about this change, which I first thought was very disruptive. Thinking it through, I don't think it's a difficult conversion and it both clears up some demonstrated strangeness in the format and enables some useful scenarios.

1. Besides enabling better search capabilities and compactness, I think this change would also address an open concern around creating deterministically ordered SARIF (to minimize code diff for log files that are used as checked in baselines). It requires specialized data structures on most platforms to deterministically order dictionary enumerations.
2. This suggestion eliminates mini languages in the format. We no longer require a convention for crafting a unique key from the uriBaseId ('#SRC_ROOT#relative/url/to/thing.cpp). We probably should have reconsidered the dictionary solution when uriBaseIds were introduced. Key collisions have always been an issue.
3. The dictionary solution was previously cited as a support for readability. Producers who are focused on readability can still produce readable files: they simply need to continue to populate the data in both the result and the referenced table. This situation is no worse than today, where that data must be duplicated (or even might exist in triplicate, consider a rule id that's in a result, used as a key, and recapitulated in the rule object). There are some minor complications locating duplicated array items that required us to define a key collision mechanism. These could be overcome if necessary by putting markers in the property bag.

If we take this change, I think we should accept it for all tables, files, logical locations, rules, etc., in order to be consistent in the format and glean benefits (such as eliminating key collision mechanism, enabling SARIF element order determinism, etc.

It should be possible to rewrite existing pre-release to account for this change.

**Comment by lgolding** [comment]
One consequence of this proposal is that `run.files` becomes `required`. Recall that at the moment, a `physicalLocation` is a `fileLocation` + a `region`, and a `fileLocation` is a `uri` + a `uriBaseId`. So you can locate the file _without_ `run.files` being present, and the only reason you need `run.files` is if you want to store _additional information_ about the file (its MIME type, its content, its hashes, _etc._).

But if `physicalLocation` contains only an index into `run.files`, then `run.files` is necessary just to locate the file.

**Comment by michaelcfanning** [comment]
I don't think this is the proposal. I think the proposal is for PLC to contain an index that can be used to reference the files array (if we convert to it).

Today I've been thinking about how/whether we could reduce key size/duplication simply by providing an alternate property for a look-up key. The ordering could be provided by a dictionary that sorts its values. So, still puzzling a bit.

**Comment by lgolding** [comment]
Oh, you're right. Scott wrote:
>  the PhysicalLocation object contains EITHER a Uri and optional UriBaseId OR an "id"

**Comment by michaelcfanning** [comment]
And there shouldn't be a problem with populating both, if you wanted the readability of a fully populated result. They key is that the index would be the only linkage to the files table. 

I'd like to hear more from Scott about how this improves searchability. There's some implied ordering in the array he proposes. Seems like this wouldn't support search in general (for example, the content isn't sorted by the MD5 hash values, for example). Perhaps he's simply saying that the more compact we can make the format, the more effective his binary JSON representation is.

If that is actually the salient point, maybe we could look into shortening the keys simply by making them an explicit property

**Comment by michaelcfanning** [comment]
We had more offline discussion. The improved searchability of this proposal likely relates to minimizing the unique set of strings in the file and putting things like results in their indexed order, to allow jumping into that table and seeking to find a result of interest.

Our current thought is to restrict this change to tables that have name collisions, these being files, logical locations and rules. Other dictionaries, e.g., resources.messageStrings, threadFlow.location.state would remain as they are.

We should review the full list of dictionary types in the format and produce the arguments on both sides for either being consistent across the format in handling them, or special-casing in order to eliminate dictionaries with synthethized keys.

**Comment by vScottLouvau** [comment]
Searchability Explanation:
- To search an array, you could simply store the position of every 128th element, say. If you were asked to find Run.Files[5138], you find the closest position you know (Run.Files[5120]) and walk forward 18 elements. If we stored 8 byte positions, we use 8 bytes / 128 files or 0.5 bits per FileData for the index.

- To search a Dictionary, you have to store every key and the position of every element. This is because the dictionary keys aren't in any guaranteed order, so if you only index some of them, you don't know which one is "closest".

In my 400MB FileData file, the memory cost to store the position of *every* 128th FileData is 79KB. The cost to store the dictionary mapping every URI to the position was ~300MB (almost four thousand times larger). I considered this unacceptable, so DeferredDictionary actually makes a Dictionary of Uri.**GetHashCode()** -> position, which reduced the usage to 33MB, at the cost of handling collisions and having to verify the URI on a seek to make sure the key really refers to the right item.

So, in summary, you can't search a dictionary without knowing about every key, but you can search an array with only a subset of items, and there are no keys themselves you must store.

**Comment by michaelcfanning** [comment]
Thanks. This is roughly the conclusion @lgolding and I reached in independent conversation. I think we're both favoring taking this change. @rtaket may have some counterbalancing thoughts.

**Comment by lgolding** [comment]
@michaelcfanning and I discussed this and here is our proposal:

### Motivation

The motivation for this proposal is to reduce memory usage (working set) for consumers of large SARIF files. Dictionary-valued properties can increase working set in three ways:

1. For dictionaries that are very large: The dictionary keys take up memory. If the keys are lengthy (for example, they are URIs), they take up even more memory.
2. For dictionaries whose keys are "synthetic" (for example, `#<uriBaseId>#<relativeReference>` in `run.files`): The dictionary keys are strings that occur nowhere else in the format, thus increasing the size of the .NET runtime's string table (or similar structure on other runtimes). Furthermore, dictionaries with synthetic keys complicate both the SARIF specification (which must explain how these keys are synthesized by producers and interpreted by consumers) and SARIF SDK implementations.

The SARIF format defines dictionaries with various combinations of these characteristics.

Replacing a dictionary with an array eliminates the memory used by the keys. It eliminates the complexity of dealing with synthetic keys. Even a naive implementation maintains fast searchability; you just have a table that contains the file offset of each element (see the original [issue description](https://github.com/oasis-tcs/sarif-spec/issues/256#issue-368433720) above). A more sophisticated implementation can save even more memory by storing only _some_ of the offsets in the lookup table -- a technique that works for arrays but does _not_ work for dictionaries (see Scott's [subsequent comment](https://github.com/oasis-tcs/sarif-spec/issues/256#issuecomment-429923116) above).

### Approach

Our approach is as follows:

- For potentially large dictionaries, replace them with arrays. In any data structure that points into these dictionaries, add a new property, an integer index into the array.
- Do the same for any other dictionaries with synthetic keys. This doesn't save much memory, but it does gain the simplification benefits of dispensing with synthetic keys.
- Leave all remaining dictionaries (those that are small and do _not_ require synthetic keys) as they are.

### Proposal
1. Change `run.files` and `run.logicalLocations` to dictionaries.

    These dictionaries are both large _and_ have synthetic keys. Add an optional `fileIndex` property to the `fileLocation` object. Add an optional `logicalLocationIndex` property to the `logicalLocation` object.

2. Change `resources.rules` to a dictionary.

    This is the only other dictionary with synthetic keys. Add an optional `ruleIndex` property to the `result` object.

3. Leave all remaining dictionaries alone:
    - `edgeTraversal.finalState`
    - `file.hashes`
    - `graphTraversal.initialState`
    - `resources.messageStrings`
    - `result.fingerprints`
    - `result.graphs`
    - `result.partialFingerprints`
    - `rule.messageStrings`
    - `rule.richMessageStrings`
    - `run.externalFiles`
    - `run.graphs`
    - `run.originalUriBaseIds`
    - `threadFlowLocation.state`
    - property bags.

**Comment by katrinaoneil** [comment]
I am assuming that Proposals 1 and 2 should actually say:

1. Change `run.files` and `run.logicalLocations` to **arrays**.

2. Change `resources.rules` to an **array**.

**Comment by lgolding** [comment]
@katrinaoneil Quite right! Sorry. Fixed.

**Comment by michaelcfanning** [comment]
fileLocation gains an index property in this proposal.
logicalLocation gains an index property
result gains a ruleIndex property.

**Comment by michaelcfanning** [comment]
consider adding count and/or startIndex when breaking these arrays apart in the external property files scenario.
